pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java

```diff
@@ -85,6 +85,7 @@
   // Keep the BrokerRequest to make incremental changes
   // TODO: Remove it once the whole query engine is using the QueryContext
   private final BrokerRequest _brokerRequest;
+  private QueryContext _preAggregateGapFillQueryContext;
```

Comment:
Can we generify this to `subquery`? I think this can be modeled as general subquery

Reply:
@Jackie-Jiang 

So this was one of the things which was discussed a lot. My concern was that changing the PinotQuery now to accommodate "**generic**" subquery has to be done carefully accounting for standard sql subquery syntax and semantics and we should be confident that it will hold in future

If we are really touching the FROM clause, my suggestion would be to make sure we understand calcite's treatment of simple FROM clause (table name as today) and complex FROM clause (sub-queries). Whatever we do today in the FROM clause to make this particular gapfill sub-query work should not interfere in the future when we are going to leverage / extend calcite's generic subquery planner to support all kinds of subqueries. 

So this is why the path of least resistance could be to not touch PinotQuery for generic Subquery support since that may require a lot of design thinking before agreeing upon how to model any subquery in Pinot. So may be making it specific like in the above code just for gapfill is the way to go ?

Reply:
@siddharthteotia @Jackie-Jiang 
Q: If we are really touching the FROM clause, my suggestion would be to make sure we understand calcite's treatment of simple FROM clause (table name as today) and complex FROM clause (sub-queries).
A: From calcite's perspective, the simple FROM clause (table name only) will be compiled as SqlIdentifier. The complex FROM clause will be compiled as SqlOrderBy object or SqlSelect object depending on if there is any order by.

It might hold true in the future to make subquery as part of DataSource instead of making the particular gapfill sub-query.

I lean towards generic approach since 
1. it is consistent with the calcite
2. even if it might impact the subquery feature, it can be fixed as part of the subquery feature development. 

@siddharthteotia any other suggestion?

Reply:
@siddharthteotia From query syntax perspective, I don't see difference between pre-aggregate-gap-fill and general subquery. Essentially the `FROM` clause contains a query instead of a table name. I'd suggest adding it as general subquery, but we can support it in the future.

Reply:
> even if it might impact the subquery feature, it can be fixed as part of the subquery feature development.

This ^^ is what I am concerned about. We are making changes to the wire object. We will have to live with it and if the decision taken now somehow does not hold for generic subquery then we will find ourselves retro-fitting subquery onto what we do now.

Note that generic subquery does not only include subquery in FROM clause but also in WHERE and other parts of a subquery. I haven't spend full time figuring out how to model all kinds of subquery in Pinot so just being extra careful here.

I think generic subquery will work and will be more cleaner. So may be let's go with it

Reply:
Done